src/command/supergraph/compose.rs

```diff
@@ -43,3 +61,197 @@ impl Compose {
         }
     }
 }
+
+pub(crate) fn get_subgraph_definitions(
+    supergraph_config: SupergraphConfig,
+    config_path: &Utf8PathBuf,
+    client_config: StudioClientConfig,
+    profile_name: &str,
+) -> Result<Vec<SubgraphDefinition>> {
+    let mut subgraphs = Vec::new();
+
+    for (subgraph_name, subgraph_data) in &supergraph_config.subgraphs {
+        match &subgraph_data.schema {
+            SchemaSource::File { file } => {
+                let relative_schema_path = match config_path.parent() {
+                    Some(parent) => {
+                        let mut schema_path = parent.to_path_buf();
+                        schema_path.push(file);
+                        schema_path
+                    }
+                    None => file.clone(),
+                };
+
+                let schema = fs::read_to_string(&relative_schema_path).map_err(|e| {
+                    let err = anyhow!("Could not read \"{}\": {}", &relative_schema_path, e);
+                    let mut err = RoverError::new(err);
+                    err.set_suggestion(Suggestion::ValidComposeFile);
+                    err
+                })?;
+
+                let url = &subgraph_data.routing_url.clone().ok_or_else(|| {
+                    let err = anyhow!("No routing_url found for schema file.");
+                    let mut err = RoverError::new(err);
+                    err.set_suggestion(Suggestion::ValidComposeRoutingUrl);
+                    err
+                })?;
+
+                let subgraph_definition = SubgraphDefinition::new(subgraph_name, url, &schema);
+                subgraphs.push(subgraph_definition);
+            }
+            SchemaSource::SubgraphIntrospection { subgraph_url } => {
+                // given a federated introspection URL, use subgraph introspect to
+                // obtain SDL and add it to subgraph_definition.
+                let client = Client::new(&subgraph_url.to_string());
+
+                let introspection_response = introspect::run(&client, &HashMap::new())?;
+                let schema = introspection_response.result;
+
+                let subgraph_definition = SubgraphDefinition::new(subgraph_name, "", &schema);
+                subgraphs.push(subgraph_definition);
+            }
+            SchemaSource::Introspection { url } => {
+                // given an introspection URL, use graph introspect to
+                // obtain SDL and add it to subgraph_definition.
+                let client = Client::new(&url.to_string());
+
+                let introspection_response = graph::introspect::run(&client, &HashMap::new())?;
+                let schema = introspection_response.result;
+
+                let subgraph_definition = SubgraphDefinition::new(subgraph_name, "", &schema);
+                subgraphs.push(subgraph_definition);
+            }
```

Comment:
Did you find a schema on which this worked?  Perhaps there's a federated implementation out there that could properly encode the nuances of federation into the introspection response, but I think the official introspection format is still lacking the depth to convey both definitions of and implementations of (on schema types) directives other than `@deprecated`.

Reply:
I was able to get it to work by just introspecting https://countries.trevorblades.com/!! o.O wild!